tests/python/contrib/test_vitis_ai_codegen.py

```diff
@@ -0,0 +1,203 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=no-else-return, unidiomatic-typecheck, invalid-name, W0611
+"""Vitis-AI codegen tests."""
+
+import numpy as np
+
+import tvm
+from tvm import relay
+from tvm.relay import transform
+from tvm.relay.op.contrib.vitis_ai import annotation
+from tvm.contrib.target import vitis_ai
+
+import pyxir
+import pyxir.contrib.target.DPUCADX8G
+
+def set_func_attr(func, compile_name, symbol_name):
+    func = func.with_attr("Primitive", tvm.tir.IntImm("int32", 1))
+    func = func.with_attr("Inline", tvm.tir.IntImm("int32", 1))
+    func = func.with_attr("Compiler", compile_name)
+    func = func.with_attr("global_symbol", symbol_name)
+    return func
+
+def _create_graph():
+    shape = (10, 10)
+    mod = tvm.IRModule()
+    x = relay.var('x', shape=shape)
+    y = relay.var('y', shape=shape)
+    z = x + x
+    p = y * y
+    func = relay.Function([x, y], p - z)
+    mod["main"] = func
+    params = {}
+    params["x"] = np.random.rand(10, 10).astype('float32')
+    params["y"] = np.random.rand(10, 10).astype('float32')
+    return mod, params
+
+
+def _construct_model(func, params=None):
```

Comment:
4 of test cases rely on this function, but this function does not check anything. It only makes sure the single op can go through the flow without crashing. I believe there are more to cover.

Reply:
Yes, that's mainly what we want to test here. We could additionally test that the operation is actually assigned for execution on the DPU accelerator. What other tests would you add?

Reply:
To be honest I actually have no concrete ideas. Usually we should test the logic of graph processing, codegen and runtime, but currently the graph processing (annotation/partition) and codegen are transparent to external tools, and we cannot really test runtime due to the lack of hardware. Regarding other related comments, it would be better if you could at least share the plan of replacing the current approach of offloading annotation/partition to XGraph with op-based annotation to give people a sense.

Reply:
These tests actually go through a lot of the interface between TVM and PyXIR, so although it doesn't contain specific checks, it thoroughly tests the interface. Also, it's actually very similar to how the CoreML codegen is tested: https://github.com/apache/incubator-tvm/blob/b8f37eeac7611e1be03adc26ee0005ddd6cdf87f/tests/python/contrib/test_coreml_codegen.py#L125 (we based ourselves on this codegen as an example). 

Regarding relying on our own partitioner and then replicating that in TVM, we do this because of the added complexity we have due to more complicated rules and constraints in the runtime for executing multiple subgraphs at the moment. We like to keep this logic in one place and be able to leverage this for integration with multiple frameworks so we currently don't have concrete plans to use the op based annotation inside TVM. Although our approach differs from the other codegens and it doesn't make much sense for us to use the op based annotation at the moment with this flow, I feel like this is actually a nice use case of BYOC as it's demonstrating the extensibility. That being said, we have some concrete plans for a new accelerator that will probably benefit a lot from the op based annotation and will look more similar to the other codegens.